advanced_source/python_custom_ops.py

```diff
@@ -0,0 +1,229 @@
+# -*- coding: utf-8 -*-
+
+"""
+Python Custom Operators
+=======================
+
+.. note::
+   This tutorial is for PyTorch 2.4+ and the PyTorch nightlies.
+
+PyTorch offers a large library of operators that work on Tensors (e.g.
+torch.add, torch.sum, etc). However, you may wish to use a new customized
+operator with PyTorch, perhaps written by a third-party library. This tutorial
+shows how to wrap Python functions so that they behave like PyTorch native
+operators. Reasons why you may wish to create a custom op in PyTorch include:
+
+- Black-box-ing an arbitrary Python function for use with torch.compile
+- Adding training support to an arbitrary Python function
+
+Please note that if your operation can be expressed as a composition of
+existing PyTorch ops, then there is usually no need to use the custom op
+API -- everything (e.g. torch.compile, training support) should just work.
+"""
+######################################################################
+# Wrapping PIL's crop into a custom op
+# ------------------------------------
+# Let's say that we are using PIL's crop operation.
+
+import torch
+from torchvision.transforms.functional import to_pil_image, pil_to_tensor
+import PIL
+import IPython
+
+def crop(pic, box):
+    img = to_pil_image(pic.cpu())
+    cropped_img = img.crop(box)
+    return pil_to_tensor(cropped_img).to(pic.device) / 255.
+
+def display(img):
+    img_pil = to_pil_image(img)
+    IPython.display.display(img_pil)
+
+
+img = torch.ones(3, 64, 64)
+img *= torch.linspace(0, 1, steps=64) * torch.linspace(0, 1, steps=64).unsqueeze(-1)
+display(img)
+cropped_img = crop(img, (10, 10, 50, 50))
+display(cropped_img)
+
+######################################################################
+# ``crop`` doesn't work performantly out-of-the-box with torch.compile. The
+# following code leads to an error when run.
+
+"""
+@torch.compile(fullgraph=True)
```

Comment:
I guess "effectively handled" above means no graph break? If so, we should state it clearly. Maybe linking to some compile doc explaining why graph breaks are not desireable